docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |   Image OK    |           0xff padding (7 octets)             ~
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    ~                       MAGIC (16 octets)                       ~
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+`Image ID:` For simplicity could be just numbering from 0 to IMAGE_CNT. It is
+used to determine which image does the status info belongs.
+
+# Proposed changes in MCUBoot
+  - Boot flow must be organized in loops which iterates over all the firmware
+    images.
+  - All image data must be stored in `boot_data` global structure.
+  - Extend global `boot_data` variable to an array and store `swap_type` per
+    image:
+``` c
+  struct boot_loader_state {
+      struct {
+          struct image_header hdr;
+          const struct flash_area *area;
+          boot_sector_t *sectors;
+          size_t num_sectors;
+      } imgs[BOOT_NUM_SLOTS];
+
+      const struct flash_area *scratch_area;
+      uint8_t swap_type;
+      uint8_t write_sz;
+  };
+
+  static struct boot_loader_state boot_data[IMAGE_CNT];
+```
+**
+  - Image trailer must be extended to record which image it belongs to.
+  - The following checks must be done in advance before any update operation to
+    ensure that only valid images will be swapped:
+      - First check for all image:
+        - `image_ok` and `copy_done` flags in image trailers in both belonging
+          slots
+        - Image integrity and authenticity
+      - Secondly check for all image, only after the checks in previous bullet
+        has finished:
+        - Dependent image version availability
+  - The final decision, which set `swap_type`, must be taken account all these
+    three checks.
+  - Image update can be started only if required swap action was determined in
+    case of all image.
+  - Setting of `copy_done` flag per image can be done only if all required image
+    update was finished. It must be a quick (should be atomic) operation, hence
+    this marks the end of the update process, only after this point can the
+    firmware images executed.
+  - In case of aborted swap the restarted boot process must make exactly the
+    same decision about required `swap_type` as the boot process did prior to
+    the aborts.
+  - Firmware image cannot be executed before swap operations has finished.
+  - Different firmware images must be signed with different keys. MCUBoot repo
+    needs to contain two key pairs, which public part must be built-in the
+    bootloader.
+  - Signature verification must be changed because signature covers the
+    dependency entries in the manifest and not just the image hash.
+
+# Proposed changes in tools
+  - Scripts has to be modified to create the dependency TLV entries in the
+    image manifest
+  - Signature generation must be changed to take into account the dependency
+    entries in the image manifest
```

Comment:
For validating this it is essential to also update the simulator to test dual image support. Once we have fully agreed on the proposal and all doubts are resolved, I can help with this part of the work.

Reply:
Thanks!